### ISSUE TITLE
Syphon Filter: Logan's Shadow  black texture when Hardware Transform on

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#7306]
openGL
Hardware Transform on
![02](https://cloud.githubusercontent.com/assets/3481559/5690437/6049e424-98c4-11e4-9580-61d96c62b4a7.png)

Hardware Transform off
![01](https://cloud.githubusercontent.com/assets/3481559/5690438/604af634-98c4-11e4-87e9-fc90ec6a74b9.png)

DX
Hardware Transform on
![03](https://cloud.githubusercontent.com/assets/3481559/5690446/95bc4bb0-98c4-11e4-83af-32151bc52296.png)

Hardware Transform off
![04](https://cloud.githubusercontent.com/assets/3481559/5690445/95b9f3ce-98c4-11e4-8d85-5ea8138c7785.png)

**Comment by daniel229** [comment]
Gedebugger

https://cloud.githubusercontent.com/assets/3481559/5690501/dddb7f5c-98c8-11e4-8393-cd46924430f5.png
https://cloud.githubusercontent.com/assets/3481559/5690498/ddd84d1e-98c8-11e4-9198-a4bdb7120539.png
https://cloud.githubusercontent.com/assets/3481559/5690499/dddad1e2-98c8-11e4-9fb2-e8edc9d3b727.png
https://cloud.githubusercontent.com/assets/3481559/5690503/dddc64a8-98c8-11e4-8cec-c7395cef03cf.png
https://cloud.githubusercontent.com/assets/3481559/5690500/dddb80ba-98c8-11e4-89b8-340456f38384.png
https://cloud.githubusercontent.com/assets/3481559/5690502/dddbf798-98c8-11e4-9846-d308359e5a39.png
https://cloud.githubusercontent.com/assets/3481559/5690504/ddff6610-98c8-11e4-8b50-89b2e1c1880f.png
https://cloud.githubusercontent.com/assets/3481559/5690506/de019fc0-98c8-11e4-8178-e2ac122c9e5a.png
https://cloud.githubusercontent.com/assets/3481559/5690505/de019110-98c8-11e4-98d1-e183281a5995.png

**Comment by unknownbrackets** [comment]
Hmm, those light types... is it not displaying them correctly?  I guess it's using point and spot.  The UVs seem odd, I wonder if it's not handling UV scale properly in the preview / vertices tab.

Hmm.

-[Unknown]

**Comment by hrydgard** [comment]
Spot is rare, we might have bugs in it..

**Comment by unknownbrackets** [comment]
Does the software renderer look correct at all?

-[Unknown]

**Comment by daniel229** [comment]
the light seems stronger in software renderer.
![01](https://cloud.githubusercontent.com/assets/3481559/5692294/58d4ac4a-992a-11e4-9256-dbec0a20086d.png)

**Comment by hrydgard** [comment]
Hm, the shader generation code is very similar between GL and DX but the results come out different. Could it be that it sends an INF or NaN in some lighting parameter, and the two shader compilers generate code that's different enough that those affect the outcome? Hm.

**Comment by DonelBueno** [comment]
You can also see by the screenshots that the light bulbs go through objects and walls in this game. It's really annoyng.

**Comment by hrydgard** [comment]
@DonelBueno , yeah, it's a typical result of the game reading the z-buffer to figure out whether to draw a light glow or not, which is hard to do correctly and quickly on desktop GL, and impossible on many OpenGL ES 2.0 implementations unfortunately - while trivial in the software renderer.

**Comment by DonelBueno** [comment]
What about OpenGL ES 3.0 and DirectX?

**Comment by hrydgard** [comment]
Well, reading back the zbuffer is very expensive on most PC and mobile GPUs that can actually do it, currently, although on beefy PCs we could afford it, it's just a bit tricky to implement, the way PPSSPP works right now. Vulkan / D3D12 might let us do things fast-and-loose like console games can, but we will see.

**Comment by unknownbrackets** [comment]
Fast and loose hurray.  Let's hope.

I don't think GLES 3.0 can do it either, without extensions (glReadPIxels can't, in GLES3, afaik anyway.)  We might be able to bind the depth as a texture and "copy" it to a color texture either via shader or what have you?  I expect this would be slow...

-[Unknown]

**Comment by hrydgard** [comment]
Well the tricky part architecture-wise right now is actually detecting that the game is accessing the zbuffer with the cpu. As VRAM is mapped like every other memory and we don't generally check addresses before r/w, we would have to resort to tricks like mapping the memory as inaccessable and trap accesses, which would trigger copying down data from the gpu. Well, of course, unless the game is nice enough to do a block copy from VRAM to RAM first where we could trivially detect the case of copying from a zbuffer, but no idea if games like this do that or not.

**Comment by unknownbrackets** [comment]
I know of at least two games that actually texture (with an incompatible format or with depal) from depth.  Talk about new complexities in depal.

Yeah, peeking into depth might even be more common than color.  But, that's not a problem we don't already have with color.

-[Unknown]

**Comment by DonelBueno** [comment]
Are there any other games affected by the lack of being able to read back from the z-buffer?

**Comment by unknownbrackets** [comment]
#5205 and #6162 at least, also some logs made me pretty sure Simpsons does something with depth as well.  Those are the ones I can remember.

-[Unknown]

**Comment by unknownbrackets** [comment]
A GE frame dump would help to confirm things here, although it seems pretty clear.

See here for instructions - it's not hard and works on Android too:
https://github.com/hrydgard/ppsspp/wiki/How-to-create-a-frame-dump

You can zip that and then drag and drop it into a reply here.

-[Unknown]

**Comment by Gamemulatorer** [comment]
There's also an issue on training 5 water
### VK & OGL
![Screenshot_2021-09-04-22-24-44-678_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/132097882-c099b7d2-95cd-4514-9d21-8dcac87fffb6.jpg)
### PSP
![Screenshot_2021-09-04-22-24-34-571_com vanced android youtube](https://user-images.githubusercontent.com/37603562/132097895-a958868f-1e7b-4cba-b0d0-a4fc46c43f1b.jpg)
```
Dump
[syphon filter.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7109920/syphon.filter.ppdmp.zip)
```

**Comment by Panderner** [comment]
> There's also an issue on training 5 water
> 
> ### VK & OGL
> ![Screenshot_2021-09-04-22-24-44-678_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/132097882-c099b7d2-95cd-4514-9d21-8dcac87fffb6.jpg)
> 
> ### PSP
> ![Screenshot_2021-09-04-22-24-34-571_com vanced android youtube](https://user-images.githubusercontent.com/37603562/132097895-a958868f-1e7b-4cba-b0d0-a4fc46c43f1b.jpg)
> Dump
> [syphon filter.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7109920/syphon.filter.ppdmp.zip)

@Gamemulatorer That bottom screenshot is not taken from the real hardware.

**Comment by Gamemulatorer** [comment]
@Panderner how did you know? 😐🤔

**Comment by Panderner** [comment]
> @Panderner how did you know? 😐🤔

Here's a video taken from real hardware:
https://youtu.be/h2Jg1BEIg7M?t=14

**Comment by Gamemulatorer** [comment]
> > @Panderner how did you know? 😐🤔
> 
> Here's a video taken from real hardware:
> https://youtu.be/h2Jg1BEIg7M?t=14

That video wasn't training 5 water.
Here's the source of my ss above https://m.youtube.com/watch?v=hTaZ2frao34&feature=youtu.be

**Comment by Panderner** [comment]
> > > @Panderner how did you know? 😐🤔
> > 
> > 
> > Here's a video taken from real hardware:
> > https://youtu.be/h2Jg1BEIg7M?t=14
> 
> That video wasn't training 5 water.
> Here's the source of my ss above https://m.youtube.com/watch?v=hTaZ2frao34&feature=youtu.be

@Gamemulatorer this is not a right footage of this video it's recorded using the emulator.

**Comment by unknownbrackets** [comment]
From that dump, this is actual rendering:

![UCUS98606_#7306_syphon_lighting](https://user-images.githubusercontent.com/191233/132105258-1565114c-d148-4c63-baae-1b1e8fcb1814.png)

Here's software, which looks pretty close except for the line on the left of the UI:

![UCUS98606_#7306_syphon_lighting_softgpu](https://user-images.githubusercontent.com/191233/132105264-2fc3a2b8-c287-4fb3-9390-cedc06135715.png)

And here's Vulkan, which looks a bit brighter (could be 32-bit vs 16-bit color):

![UCUS98606_#7306_syphon_lighting_vulkan](https://user-images.githubusercontent.com/191233/132105279-8c31a8ca-b2e7-4e2a-8155-51a9a2d47cbe.png)

In each case the light looks about the same.

Should we mark this issue as fixed?  It could've been fixed by several other lighting fixes, such as the one that fixed Virtua Tennis.

-[Unknown]

**Comment by hrydgard** [comment]
Pretty sure both issues here have been fixed. Closing.
